procedures/statement_performance_analyzer.sql

```diff
@@ -0,0 +1,792 @@
+        IF (NOT INSTR(in_table, '.')) THEN
+            -- No . in the table name - use current database
+            -- DATABASE() will be the database of the procedure
+            SET v_table_db   = DATABASE(),
+                v_table_name = in_table;
+        ELSE
+            SET v_table_db   = SUBSTRING_INDEX(in_table, '.', 1);
+            SET v_table_name = SUBSTRING(in_table, CHAR_LENGTH(v_table_db)+2);
+        END IF;
+
+        SET v_quoted_table = CONCAT('`', v_table_db, '`.`', v_table_name, '`');
+
+        IF (@sys.debug = 'ON') THEN
+            SELECT CONCAT('in_table is: db = ''', v_table_db, ''', table = ''', v_table_name, '''') AS 'Debug';
+        END IF;
+
+        IF (v_table_db = DATABASE() AND (v_table_name = 'tmp_digests' OR v_table_name = 'tmp_digests_delta')) THEN
+            SET v_error_msg = CONCAT('Invalid value for in_table: ', v_quoted_table, ' is reserved table name.');
+            SIGNAL SQLSTATE '45000'
+               SET MESSAGE_TEXT = v_error_msg;
+        END IF;
+
+        CALL table_exists(v_table_db, v_table_name, v_table_exists);
+        IF (@sys.debug = 'ON') THEN
+            SELECT CONCAT('v_table_exists = ', v_table_exists) AS 'Debug';
+        END IF;
+
+        IF (v_table_exists IN ('BASE TABLE', 'TEMPORARY')) THEN
+            -- Verify that the table has the correct table definition
+            -- This also minimises the risk of using a production table.
+            SET v_checksum_ref = (
+                 SELECT GROUP_CONCAT(CONCAT(COLUMN_NAME, COLUMN_TYPE) ORDER BY ORDINAL_POSITION) AS Checksum
+                   FROM information_schema.COLUMNS
+                  WHERE TABLE_SCHEMA = 'performance_schema' AND TABLE_NAME = 'events_statements_summary_by_digest'
+                ),
+                v_checksum_table = (
+                 SELECT GROUP_CONCAT(CONCAT(COLUMN_NAME, COLUMN_TYPE) ORDER BY ORDINAL_POSITION) AS Checksum
+                   FROM information_schema.COLUMNS
+                  WHERE TABLE_SCHEMA = v_table_db AND TABLE_NAME = v_table_name
+                );
+
+            IF (v_checksum_ref <> v_checksum_table) THEN
+                -- The table does not have the correct definition, so abandon
+                SET v_error_msg = CONCAT('The table ',
+                                         IF(CHAR_LENGTH(v_quoted_table) > 93, CONCAT('...', SUBSTRING(v_quoted_table, -90)), v_quoted_table),
+                                         ' has the wrong definition.');
+                SIGNAL SQLSTATE '45000'
+                   SET MESSAGE_TEXT = v_error_msg;
+            END IF;
+        END IF;
+    END IF;
+
+
+    IF (in_views IS NULL OR in_views = '') THEN
+        -- Set to default
+        SET in_views = 'with_runtimes_in_95th_percentile,analysis,with_errors_or_warnings,with_full_table_scans,with_sorting,with_temp_tables';
+    END IF;
+
+
+    -- Validate settings
+    CALL table_exists(DATABASE(), 'tmp_digests', v_tmp_digests_table_exists);
```

Comment:
Prefix this with "sys.table_exists", these can be called from outside of the sys schema.